### ISSUE TITLE
머딥 - 7.5d (5~7 복습 정리)

### ISSUE THREAD

**Issue by vmtmxmf5** [vmtmxmf5/Python-Basics#35]
# 머신러닝 복습용 A to Z

## 데이터 로드
```sage
import os
import numpy as np
import pandas as pd

HOUSING_PATH = os.path.join('datasets', 'housing')

def load_housing_data(housing_path=HOUSING_PATH, filename="housing.csv"):
  csv_path = os.path.join(housing_path, filename)
  return pd.read_csv(csv_path)

housing = load_housing_data()
```

## stratified train/test set
```sage
housing['income_cat'] = pd.cut(
    housing['median_income'],
    bins=[0, 1.5, 3, 4.5, 6, np.inf],
    labels=[1, 2, 3, 4, 5])

from sklearn.model_selection import StratifiedShuffleSplit

strat_shuff_split = StratifiedShuffleSplit(
    n_splits=1,
    test_size=0.2,
    random_state=42
    )

for train_idx, test_idx in strat_shuff_split.split(housing, housing['income_cat']):
    strat_train_set = housing.loc[train_idx]
    strat_test_set = housing.loc[test_idx]
    
for set_ in (strat_train_set, strat_test_set):
    set_.drop('income_cat', axis=1, inplace=True)
    
housing = strat_train_set.copy()

label_name = 'median_house_value'
housing_labels = strat_train_set[label_name].copy()
housing = strat_train_set.drop(label_name, axis=1)
```

## Pipeline
```sage
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import OneHotEncoder
from sklearn.base import BaseEstimator, TransformerMixin

housing_num = housing.drop('ocean_proximity', axis=1)
housing_cat = housing[['ocean_proximity']]

rooms_ix, bedrooms_ix, population_ix, households_ix = 3, 4, 5, 6
class CombinedAttributeAdder(BaseEstimator, TransformerMixin):
    def fit(self, X):
        return self
    
    def transform(self, X):
        rooms_per_households = X[:, rooms_ix] / X[:, households_ix]
        bedrooms_per_rooms = X[:, bedrooms_ix] / X[:, rooms_ix]
        return np.c_[X, rooms_per_households, bedrooms_per_rooms]
    
attr_adder = CombinedAttributeAdder()
housing_extra_attribs = attr_adder.transform(housing.values)


from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler
from sklearn.compose import ColumnTransformer

num_pipeline = Pipeline([
    ('imputer', SimpleImputer(strategy='median')),
    ('add_attr', CombinedAttributeAdder()),
    ('std_scaler', StandardScaler())
    ])

num_feature_name = list(housing_num.columns)
cat_feature_name = ['ocean_proximity']

full_pipeline = ColumnTransformer([
    ('num', num_pipeline, num_feature_name),
    ('cat', OneHotEncoder(), cat_feature_name)
    ])

housing_final = full_pipeline.fit_transform(housing)
```

## 모델링
```sage
from sklearn.linear_model import LinearRegression
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error

lin_reg = LinearRegression()
lin_reg.fit(housing_final, housing_labels)

lin_pred = lin_reg.predict(housing_final)
lin_mse = mean_squared_error(housing_labels, lin_pred)
lin_rmse = np.sqrt(lin_mse)


tree_reg = DecisionTreeRegressor()
tree_reg.fit(housing_final, housing_labels)

tree_pred = tree_reg.predict(housing_final)
tree_mse = mean_squared_error(housing_labels, tree_pred)
tree_rmse = np.sqrt(tree_mse)


forest_reg = RandomForestRegressor()
forest_reg.fit(housing_final, housing_labels)

forest_pred = forest_reg.predict(housing_final)
forest_mse = mean_squared_error(housing_labels, forest_pred)
forest_rmse = np.sqrt(forest_mse)
```

## cross-validation (한방에 여러 모델 하려면 앙상블 모델 ㄱㄱ)
```sage
from sklearn.model_selection import cross_val_score

lin_score = cross_val_score(
    LinearRegression(), 
    housing_final,
    housing_labels,
    cv=10,
    scoring='neg_mean_squared_error'
    )

lin_rmse_score = np.sqrt(-lin_score)
```

## hyper parameter
```sage
from sklearn.model_selection import GridSearchCV

params = {
    'n_estimators':[5, 10, 15],
    'max_features':[4, 7, 10]
    }

grid_search = GridSearchCV(
    RandomForestRegressor(),
    params,
    cv=10,
    scoring='neg_mean_squared_error',
    return_train_score=True
    )

grid_search.fit(housing_final, housing_labels)
final_model = grid_search.best_estimator_
```

## test셋 준비하기
```sage
X_test = strat_test_set.drop('median_house_value', axis=1)
y_test = strat_test_set['median_house_value'].copy()

X_test_final = full_pipeline.transform(X_test)

final_pred = final_model.predict(X_test_final)

final_mse = mean_squared_error(y_test, final_pred)
final_rmse = np.sqrt(final_mse)

print(final_rmse)
```
